cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        """
+        status = await async_get_status()
+        app_status = status.applications.get(self.name)
+        units_not_upgraded = []
+        for unit in app_status.units.keys():
+            workload_version = app_status.units[unit].workload_version
+            compatible_os_versions = OpenStackCodenameLookup.lookup(self.charm, workload_version)
+            if target not in compatible_os_versions:
+                units_not_upgraded.append(unit)
+        if units_not_upgraded:
+            logger.error(
+                "App: '%s' has units: '%s' didn't upgrade to %s",
+                self.name,
+                ", ".join(units_not_upgraded),
+                str(target),
+            )
+
+    def pre_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Pre Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add pre upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        return [self._get_refresh_charm_plan(target)]
+
+    def upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :raises HaltUpgradePlanGeneration: When the application halt the upgrade plan generation
+        :return: Plan that will add upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        if self.current_os_release >= target:
+            logger.warning(
+                "Application: '%s' already on a newer version than %s. Ignoring.",
+                self.name,
+                target,
+            )
+            raise HaltUpgradePlanGeneration()
+        return [
+            self._get_disable_action_managed_plan(),
+            self._get_upgrade_charm_plan(target),
+            self._get_workload_upgrade_plan(target),
+        ]
+
+    def post_upgrade_plan(self, target: OpenStackRelease) -> list[UpgradeStep]:
+        """Post Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add post upgrade as sub steps.
+        :rtype: list[UpgradeStep]
+        """
+        return [self._get_reached_expected_target_plan(target)]
+
+    def generate_upgrade_plan(self, target: str) -> UpgradeStep:
+        """Generate full upgrade plan for an Application.
+
+        :param target: OpenStack codename to upgrade.
+        :type target: str
+        :return: Full upgrade plan if the Application is able to generate it.
+        :rtype: UpgradeStep
+        """
+        target_version = OpenStackRelease(target)
+        upgrade_plan = UpgradeStep(
```

Comment:
nit: you already have an `upgrade_plan` function. Renaming this variable to a more unique name (e.g. `upgrade_step`) would make reading and searching through the code easier

Reply:
Done